UnityProject/Assets/Scripts/Items/Tool/AirlockPainter.cs

```diff
@@ -45,16 +58,29 @@ public void OnSpawnServer(SpawnInfo info)
 			Inventory.ServerSpawnPrefab(tonerPrefab, tonerSlot);
 		}
 
-		public async void ChoosePainJob(GameObject performer)
+		public void ChoosePaintJob()
+		{
+			if (AvailablePaintJobs is null || RadialOptions is null) return;
+
+			var controller = RightClickManager.Instance.MenuController;
+			controller.SetupMenu(PainterMenuItems, RadialScreenPosition, RadialOptions);
+		}
+
+		public List<RightClickMenuItem> GeneratePaintMenu(List<GameObject> objects)
 		{
-			if (AvailablePaintJobs == null) return;
+			if (objects is null || objects.Count == 0) return null;
 
-			GameObject chosenPaintJob = await UIManager.RadialMenu.ShowRadialMenu(AvailablePaintJobs, performer);
-			if (chosenPaintJob)
+			var result = new List<RightClickMenuItem>();
+
+			for (var i = 0; i < objects.Count; i++)
 			{
-				int chosenPaintJobIndex = AvailablePaintJobs.IndexOf(chosenPaintJob);
-				PlayerManager.LocalPlayerScript.playerNetworkActions.CmdSetPaintJob(chosenPaintJobIndex);
+				var index = i; // Copy for action function
+				Action setPaint = () => PlayerManager.LocalPlayerScript.playerNetworkActions.CmdSetPaintJob(index);
```

Comment:
I don't get it, why couldn't you just use 'i'?

Reply:
Had a feeling that would come up. TLDR: If 'i' was used, all the actions would choose the last airlock in the list.

The reason has to do with the way variables are captured in anonymous functions/lambdas. The lambdas would capture a reference to 'i' and all the actions would use that reference. Once the loop finishes, 'i' will be the final value in the list. This means all the setPaint actions would be say, 10, after the loop runs. Instead, we create a new variable 'index' on each iteration and copy the value of 'i' into it. Then the setPaint lambdas capture the individual 'index' references which have the values we want (0, 1, 2... etc).

A foreach doesn't have the same issue with capturing variables as a for loop does. The loop variable is created on each iteration in foreach. I could use a foreach using objects.Select((obj, i) => i) but I figured I would get a similar response ("why not just use a for loop"). I might be able to use a foreach with Enumerable.Range if it's available in the current version of unity. I'll check if unity has it when I get a chance.

Reply:
nah, the for loop is fine. I see why you did it now.